### PR TITLE
[CARBONDATA-3196] [CARBONDATA-3203]Fixed Compaction for Complex types with Dictionary Include and also supported Compaction for restructured table

### PR DESCRIPTION
**Pull request by manishnalla1994**
Problem1: Compaction Failing for Complex datatypes with Dictionary Include as KeyGenenrator was not being set in model for Dictionary Include Complex Columns and dictionary include complex columns were not handled for finding cardinality.

Solution: Handled both these issues by setting KeyGenerator and storing cardinality of Complex dictionary include columns.

Problem2: Compaction was failing for restructured table containing dictionary include complex columns.

Solution: Handled complex columns for this case by inserting correct indices of the columns.

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [x] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1929/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1930/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10183/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2140/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1952/

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/complexType/TestCompactionComplexType.scala [hunk 1]
```diff
@@ -1068,4 +1068,40 @@ class TestCompactionComplexType extends QueryTest with BeforeAndAfterAll {
     sql("Drop table if exists adaptive")
   }
 
+  test("Test major compaction for struct of array type") {
+    sql("DROP TABLE IF EXISTS carbon")
```
Comment (manishgupta88) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/complexType/TestCompactionComplexType.scala]: 1. Include dropping of table in afterAll also
2. Give the test case description as below
`Test major compaction with dictionary include for struct of array type`

File: processing/src/main/java/org/apache/carbondata/processing/store/CarbonFactDataHandlerModel.java [hunk 2]
```diff
@@ -356,9 +359,20 @@ public static CarbonFactDataHandlerModel getCarbonFactDataHandlerModel(CarbonLoa
...
-    // get the cardinality for all all the columns including no dictionary columns
-    int[] formattedCardinality = CarbonUtil
-        .getFormattedCardinality(segmentProperties.getDimColumnsCardinality(), wrapperColumnSchema);
+    // get the cardinality for all all the columns including no
+    // dictionary columns and complex columns
+    int[] dimAndComplexColumnCardinality =
+        new int[segmentProperties.getDimColumnsCardinality().length + segmentProperties
+            .getComplexDimColumnCardinality().length];
+    for (int i = 0; i < segmentProperties.getDimColumnsCardinality().length; i++) {
+      dimAndComplexColumnCardinality[i] = segmentProperties.getDimColumnsCardinality()[i];
```
Comment (manishgupta88) [inline processing/src/main/java/org/apache/carbondata/processing/store/CarbonFactDataHandlerModel.java]: Check for a resturcture drop column case when few loads are done and then dictionary column is dropped and compaction is triggered. In that case segmentProperties will contain cardinality of dropped column also check finally what is the schema and cardinality written during compaction

File: processing/src/main/java/org/apache/carbondata/processing/util/CarbonDataProcessorUtil.java [hunk 3]
```diff
@@ -699,4 +701,33 @@ public static boolean isRawDataRequired(CarbonDataLoadConfiguration configuratio
     return iterators;
   }
 
+  public static int[] calcDimensionLengths(int numberOfSortColumns, int[] complexCardinality) {
+    if (!(numberOfSortColumns > 0)) {
```
Comment (manishgupta88) [inline processing/src/main/java/org/apache/carbondata/processing/util/CarbonDataProcessorUtil.java]: 1. Rewrite the condition `if (!(numberOfSortColumns > 0))` as `if (numberOfSortColumns == 0)`
2. The functionality of this method is not clear. Add a comment to explain the logic explanation and use of this method

File: processing/src/main/java/org/apache/carbondata/processing/util/CarbonDataProcessorUtil.java [hunk 4]
```diff
@@ -699,4 +701,33 @@ public static boolean isRawDataRequired(CarbonDataLoadConfiguration configuratio
...
+        if (complexCardinality[i] != 0) {
+          complexCardinality[i] = Integer.MAX_VALUE;
+        }
+      }
+    }
+    List<Integer> dimsLenList = new ArrayList<Integer>();
+    for (int eachDimLen : complexCardinality) {
+      if (eachDimLen != 0) dimsLenList.add(eachDimLen);
+    }
+    int[] dimLens = new int[dimsLenList.size()];
```
Comment (manishgupta88) [inline processing/src/main/java/org/apache/carbondata/processing/util/CarbonDataProcessorUtil.java]: Conversion from arrayList to array can be done directly

File: processing/src/main/java/org/apache/carbondata/processing/util/CarbonDataProcessorUtil.java [hunk 5]
```diff
@@ -699,4 +701,33 @@ public static boolean isRawDataRequired(CarbonDataLoadConfiguration configuratio
...
+    for (int i = 0; i < dimsLenList.size(); i++) {
+      dimLens[i] = dimsLenList.get(i);
+    }
+    return dimLens;
+  }
+
+  public static KeyGenerator[] createKeyGeneratorForComplexDimension(int numberOfSortColumns,
```
Comment (manishgupta88) [inline processing/src/main/java/org/apache/carbondata/processing/util/CarbonDataProcessorUtil.java]: Add a method comment to explain the logic and method usage

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10205/

File: processing/src/main/java/org/apache/carbondata/processing/store/CarbonFactDataHandlerModel.java (hunk 2 above)
Comment (manishnalla1994) [reply processing/src/main/java/org/apache/carbondata/processing/store/CarbonFactDataHandlerModel.java]: The restructure case is not handled for complex types compaction. I have raised the JIRA issue and I will handle it. Please find the JIRA link : 'https://issues.apache.org/jira/browse/CARBONDATA-3203' .

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2227/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2257/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2051/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10303/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2062/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2266/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10315/

**Comment by manishgupta88** [comment]
LGTM

**Pull request closed without merging**